src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -941,8 +941,9 @@
   (if (or (nil? expr)
           (not (subquery-expression? expr)))
     (let [qr (assoc query-rec :where (user-node->plan-node query-rec expr))]
-      (if (:late-project? query-rec)
-        (assoc qr :late-projected-fields column-list)
+      (if-let [must-haves (seq (keys (filter (comp :must-have? val) (:projections query-rec))))]
```

Comment:
This line is bugging me... I think it's probably because I don't understand the structure of `query-rec`, so I couldn't really tell you what it's doing. This could be a good place for an internal schema, if indeed they all share a schema. 

(I'm not asking for a change here, just thinking aloud)